devdocs/PIPELINE_REFACTORING_PLAN.md

```diff
@@ -2,276 +2,481 @@
+   - Takes configuration, collection metadata, and available files
+   - Returns a complete specification of geometry decisions
+   - Contains all business rules for geometry selection
+4. **Document decision rules** clearly:
+   - Priority order for data sources
+   - Fallback logic
+   - Override conditions
+5. **Add `GeometrySpec` to pipeline context** for use in later phases
+
+**Benefit if stopped here**: Geometry decision logic is isolated from execution, making it testable and comprehensible
+
+### Phase 3: Temporal Specification Extraction
+**Standalone Value**: Clean separation of temporal decision-making from execution
+
+1. **Extract temporal decision logic** from current pipeline
+2. **Define `TemporalSpec` dataclass** to capture decisions:
+   - Source: where to get temporal data (granule metadata, premet file, collection)
+   - Type: what temporal type to create (single datetime, range datetime)
+3. **Create `determine_temporal_spec()` function** that:
+   - Takes configuration, collection metadata, and available files
+   - Returns a complete specification of temporal decisions
+   - Encapsulates all business rules for temporal selection
+4. **Document decision rules** clearly:
+   - When to use single vs range datetime
+   - Collection temporal override logic
+   - Validation requirements
+5. **Add `TemporalSpec` to pipeline context** for use in later phases
+
+**Benefit if stopped here**: Temporal decision logic is isolated from execution, testable independently
+
+### Phase 4: Pipeline State Creation
+**Standalone Value**: Organized pipeline with clear data flow
+
+1. **Create `PipelineState`** class containing:
+   - Configuration
+   - CMR metadata (from Phase 1)
+   - Geometry specification (from Phase 2)
+   - Temporal specification (from Phase 3)
+   - Reader registry (empty for now)
+   - Processing ledger
+2. **Create new `Pipeline` class** that uses state
+3. **Update pipeline operations** to use specifications:
+   - Operations read the specs to know what to do
+   - Specs tell WHERE to get data and WHAT type to create
+   - Operations handle the HOW of extraction and formatting
+4. **Ensure all actions are pure functions** that return new state
+
+**Benefit if stopped here**: Clear separation between decision-making (specs) and execution (operations)
+
+### Phase 5: Reader Interface Definition
+**Standalone Value**: Standardized interface for all metadata extraction
+
+1. **Define new granular reader interfaces**:
+   - `TemporalReader`: Extracts temporal metadata
+   - `SpatialReader`: Extracts spatial metadata
+   - `AttributeReader`: Extracts general metadata/attributes
+2. **Create `ReaderRegistry`** to manage reader selection
+3. **Define reader selection logic** based on file types
+4. **Update pipeline state** to include reader registry
+
+**Benefit if stopped here**: Clear contract for all readers, foundation for extensibility
+
+### Phase 6: Custom Reader Implementation
```

Comment:
I strongly feel that the project would benefit from putting what you currently have in Phase 7 here, and delaying the custom reader implementation to later. The "custom reader" idea is sexy, but we can handle premets and spatial files and people can use their custom code to generate those. Refactoring our existing readers to fit the new architecture will make maintenance of the entire application easier down the road.